### PR TITLE
Improve clean/reset of namespaces

### PR DESCRIPTION
**Pull request by karlkfi**
- Use server-side filter to avoid deleting shared namespaces.
- Make it more clear which namespaces are protected and which are
  shared between tests.
- Use client-go to delete namespaces in clean, instead fo kubectl

### CODE REVIEW

File: e2e/nomostest/reset.go [hunk 1]
```diff
@@ -441,6 +454,19 @@ func withLabelListOption(key, value string) client.MatchingLabelsSelector {
 	return client.MatchingLabelsSelector{Selector: labelSelector}
 }
 
+func withNameNotInListOption(values ...string) client.MatchingFieldsSelector {
```
Comment (sdowell) [inline e2e/nomostest/reset.go:457]: Looks like the input arg is unused here and it's always using `sharedTestNamespaces`. Should update this to use the input arg

File: e2e/nomostest/reset.go [hunk 2]
```diff
@@ -39,6 +40,32 @@ import (
...
+	testGitNamespace,
+}
+
+// protectedNamespaces is a list of namespaces that should never be deleted.
+//
+// Reset will error if these namespaces exist and have the test label.
+//
+// Individual test cleanup MUST revert these namespaces, if modified.
+// See `checkpointProtectedNamespace` for an example.
+var protectedNamespaces = func() []string {
```
Comment (sdowell) [inline e2e/nomostest/reset.go:60]: I didn't mind having these arrays returned by a function as it ensures you always get a pseudo-immutable value (even if it's a bit suboptimal), but I don't feel too strongly about it

File: e2e/nomostest/clean.go [hunk 3]
```diff
@@ -218,15 +220,66 @@ func Clean(nt *NT, failOnError FailOnError) {
...
+	return DeleteNamespaces(nt, nsListItems)
+}
+
+// DeleteNamespaces cleans up one or more Namespaces and all their namespaced objects.
+// Use this for bulk deletion of Namespaces.
+func DeleteNamespaces(nt *NT, nsList []corev1.Namespace) error {
+	if len(nsList) == 0 {
+		return nil
+	}
+	nt.T.Logf("[CLEANUP] Deleting Namespaces (%d)", len(nsList))
```
Comment (sdowell) [inline e2e/nomostest/clean.go:264]: Consider moving this line above the zero length check so we know if the array was zero length from the test logs

File: e2e/nomostest/reset.go (hunk 2 above)
Comment (karlkfi) [reply e2e/nomostest/reset.go:60]: I suspect it's less important for a private variable where you control all usages. But yeah, it's sort of a language failing that there's no immutable collections you can use as constant.

Java users fixed this problem by adding libraries for ImmutableList and such, but that's hard to do in Go where slices and arrays aren't structs and you can't override their methods or provide alternate impls for their interfaces.

File: e2e/nomostest/reset.go (hunk 1 above)
Comment (karlkfi) [reply e2e/nomostest/reset.go:457]: good catch

File: e2e/nomostest/clean.go (hunk 3 above)
Comment (karlkfi) [reply e2e/nomostest/clean.go]: Done

File: e2e/nomostest/reset.go (hunk 1 above)
Comment (karlkfi) [reply e2e/nomostest/reset.go:428]: Fixed

**Pull request merged**